### PR TITLE
Issue #31: improve agasc tests

### LINKED ISSUE #31: Idea for agasc updates: testing and better API and legacy compatibility
This is an idea for consideration so we can promote Ska3 without a known set of test fails for `agasc` causing a red mark on our ska_testr results.

It seems we need a new release of Ska3-flight and Ska3-matlab anyway for the mica dark cal file handle leak problem, here is an idea for agasc improvements that Malgosia could lead:
- Modify the test mp_get_agasc() function to use a fixed radius of 0.6 deg and to write out the results to a gzipped FITS file.  These will become the new reference standard and will be included in the package distribution.  Label the files in a way that includes the info that they are from AGASC 1.6.
- Now modify the tests to use a radius=0.5 deg in testing and use those pre-generated files and compare with results for agasc_1p6.h5.
- Remove all random elements from testing.  It should be fixed and deterministic.
- You'll need to bundle package data in the setup.py.  See https://github.com/sot/proseco/blob/master/setup.py#L17 for an example (but the data should be in `agasc.tests`).
- Update the code so that if `agasc_file` is provided that is not an absolute file, then it looks for `Path(os.environ['SKA'], 'data', 'agasc', agasc_file)`.  If not provided the default should be ~~'miniagasc_1p7.h5'~~ 'miniagasc.h5'.  (Use `Path` everywhere, not `os.path`).
- I think this can be a Ska3-only update, so update the version to 4.6.
- ~~When this is installed and starcheck-py3 is promoted, we could optionally revert the miniagasc.h5 link to point back at agasc_1p6.h5 for back-compatibility with legacy code in Ska2.~~

Thoughts?  We can limp through the finish line without this, but it would be nicer with something like this.

### PR DESCRIPTION
**Pull request by malgosias**
This is WIP on improving agasc tests. Main changes include:
- **mp_get_agasc** created fits file for a set of (ra, dec, radius) parameters. If the files exist the data is read from these files. See also setup.py for package_data definition. **Issue:** I understood that in the final version we do not  want to write fits files, so in the next step this needs to be updated so that we only read from the package_data files. **Related issue:** do we still want to keep interactive_test_agasc? If yes, then in the current version of mp_get_agasc I would only comment out the line that writes the fits file.
- AGASC_FILE is defined explicitly to point to miniagasc_1p6.h5
-  removed random numbers from the definition of ras and decs 

I confirmed that with this implementation all tests pass (on my desktop, where ascds environment is available), except of interactive_test_agasc, which I didn't run because of the random factor.

Let me know if this what you had in mind and how to proceed.

Closes #31.

### CODE REVIEW

**Comment by taldcroft** [comment]
@malgosias - I tried to test and found some issues, so I just finished this PR up this morning.

Some notes:
- I had suspected the files were getting written into the wrong directory from looking at the code, but I'm always confused about the current working directory within pytest.  Anyway, for future reference you need to be explicit about the path of files.
- My comment about testing the packaging was key and saved myself from a mistake fixed in 
5228e71.  Also your original commits were not including the FITS files into the git repo.

**Comment by taldcroft** [comment]
Testing:
- [x] `python setup.py test` on ska3 linux-64 (kadi), ska3 linux-32 (quango), ska3 macosx (laptop)
- [x] Install as a package (`python setup.py sdist; pip install dist/agasc-4.6.tar.gz`) and pass tests from ipython.
- [x] Testing of the AGASC ID (test_agasc_id) is always skipping because I don't have ksh.

**Comment by taldcroft** [comment]
On reflection this misses a key goal of testing on Windows.  Stand by.

File: agasc/agasc.py [hunk 1]
```diff
@@ -25,6 +24,16 @@ class InconsistentCatalogError(Exception):
     pass
 
 
+def tables_open_file(filename):
```
[the remark this answers is not in the document]
Comment (taldcroft) [reply agasc/agasc.py]: Needed for the `str()` wrapper inside to handle `Path` objects.

**Comment by taldcroft** [comment]
The testing checklist from above still passes (with expected skips), but now with more tests.

The new commit now generates the reference data from the miniagasc if it is not available and requested.  Note that the ref data in the repo is for 1p6 is from the DS AGASC (mp_get_agasc) while for 1p7 it is from miniagasc_1p7.  In the end it is all identical, so should be OK.

The new commits now check what `get_agasc_cone` fetches vs. the DS AGASC if that is available.  Going forward reference data will always be generated from our HDF5 files, so will serve as regression test values.  On platforms where DS is available then consistency with the DS AGASC will be checked.

Whew!

**Comment by jeanconn** [comment]
Oh, I'm guessing that the tests are fine on non-HEAD systems, but that test_agasc_id hasn't been updated.   Should the test be removed as OBE or?

**Comment by taldcroft** [comment]
OK, the issue is that kadi does not have `ksh` so that test was being skipped.  But as advertised I did run the tests on ska3/flight on kadi, quango, and did not see any fails.  I'll fix and test on fido.

**Comment by taldcroft** [comment]
OK, passing all tests (no skips) on fido /proj/sot/ska3/flight, and passing tests with skips on quango /proj/sot/ska3/matlab32, and Mac ska3-flight:
```
ska3-fido$ python setup.py test --args='-v'
running test
running egg_info
writing agasc.egg-info/PKG-INFO
writing dependency_links to agasc.egg-info/dependency_links.txt
writing top-level names to agasc.egg-info/top_level.txt
reading manifest file 'agasc.egg-info/SOURCES.txt'
writing manifest file 'agasc.egg-info/SOURCES.txt'
running build_ext
==================================================================== test session starts ====================================================================
platform linux -- Python 3.6.2, pytest-3.2.1, py-1.4.34, pluggy-0.4.0 -- /proj/sot/ska3/flight/bin/python
cachedir: .cache
rootdir: /data/baffin/tmp/agasc, inifile:
plugins: remotedata-0.3.0, openfiles-0.3.0, doctestplus-0.1.3, arraydiff-0.2
collected 24 items                                                                                                                                           

agasc/tests/test_agasc_1.py::test_multi_agasc PASSED
agasc/tests/test_agasc_2.py::test_agasc_conesearch[0.0-89.9-1p6] PASSED
agasc/tests/test_agasc_2.py::test_agasc_conesearch[0.0-89.9-1p7] PASSED
agasc/tests/test_agasc_2.py::test_agasc_conesearch[180.0--89.9-1p6] PASSED
agasc/tests/test_agasc_2.py::test_agasc_conesearch[180.0--89.9-1p7] PASSED
agasc/tests/test_agasc_2.py::test_agasc_conesearch[0.1-0.0-1p6] PASSED
agasc/tests/test_agasc_2.py::test_agasc_conesearch[0.1-0.0-1p7] PASSED
agasc/tests/test_agasc_2.py::test_agasc_conesearch[180.0-0.0-1p6] PASSED
agasc/tests/test_agasc_2.py::test_agasc_conesearch[180.0-0.0-1p7] PASSED
agasc/tests/test_agasc_2.py::test_agasc_conesearch[275.36-8.09-1p6] PASSED
agasc/tests/test_agasc_2.py::test_agasc_conesearch[275.36-8.09-1p7] PASSED
agasc/tests/test_agasc_2.py::test_against_ds_agasc[0.0-89.9] PASSED
agasc/tests/test_agasc_2.py::test_against_ds_agasc[180.0--89.9] PASSED
agasc/tests/test_agasc_2.py::test_against_ds_agasc[0.1-0.0] PASSED
agasc/tests/test_agasc_2.py::test_against_ds_agasc[180.0-0.0] PASSED
agasc/tests/test_agasc_2.py::test_against_ds_agasc[275.36-8.09] PASSED
agasc/tests/test_agasc_2.py::test_basic PASSED
agasc/tests/test_agasc_2.py::test_float16 PASSED
agasc/tests/test_agasc_2.py::test_proper_motion PASSED
agasc/tests/test_agasc_2.py::test_add_pmcorr_is_consistent[1180612288-2020:001-219.864331--60.831868-high proper motion, epoch 2000] PASSED
agasc/tests/test_agasc_2.py::test_add_pmcorr_is_consistent[198451217-2020:001-247.892206-19.276605-epoch 1982 star] PASSED
agasc/tests/test_agasc_2.py::test_add_pmcorr_is_consistent[501219465-2020:001-166.998976-52.82208-epoch 1984 star] PASSED
agasc/tests/test_agasc_2.py::test_agasc_id[0.0-89.9] PASSED
agasc/tests/test_agasc_2.py::test_agasc_id[180.0--89.9] PASSED
```

**Pull request merged**